lib/node.js

```diff
@@ -533,11 +580,492 @@ class Node {
+   * 
+   * If an error occurs, the process is stopped and returned. Parameters
+   * set before an error remain unchanged.
+   * 
+   * @param {Parameter[]} parameters - The parameters to set.
+   * @return {rclnodejs.rcl_interfaces.msg.SetParameterResult[]} - A list of SetParameterResult, one for each parameter that was set.
+   */
+  setParameters(parameters = []) {
+    return parameters.map(parameter =>
+      this.setParametersAtomically([parameter])
+    );
+  };
+
+  /**
+   * Repalce a list of declared parameters atomically.
+   * 
+   * Declared parameters are replaced in the order they are provided.
+   * A single ParameterEvent is published collectively for all changed
+   * parameters.
+   * 
+   * If an error occurs, the process stops immediately. All parameters updated to 
+   * the point of the error are reverted to their previous state.
+   * 
+   * @param {Parameter[]} parameters - The parameters to set.
+   * @return {rclnodejs.rcl_interfaces.msg.SetParameterResult} - describes the result of setting 1 or more parameters.
+   */
+  setParametersAtomically(parameters = []) {
+    return this._setParametersAtomically(parameters);
+  };
+
+  /**
+   * Internal method for updating parameters atomically.
+   * 
+   * @param {Paramerter[]} parameters - The parameters to update.
+   * @param {boolean} declareParameterMode - When true parameters are being declared; 
+   *    otherwise they are being changed.
+   * @return {SetParameterResult} - A single collective result.
+   */
+  _setParametersAtomically(parameters = [], declareParameterMode = false) {
+    let result = this._validateParameters(parameters, declareParameterMode);
+    if (!result.successful) {
+      return result;
+    }
+
+    // collectively track updates to parameters for use 
+    // when publishing a ParameterEvent
+    const newParameters = [];
+    const changedParameters = [];
+    const deletedParameters = [];
+
+    for (const parameter of parameters) {
+      if (parameter.type == ParameterType.PARAMETER_NOT_SET) {
+        this.undeclareParameter(parameter.name);
+        deletedParameters.push(parameter);
+      } else {
+        this._parameters.set(parameter.name, parameter);
+        if (declareParameterMode) {
+          newParameters.push(parameter);
+        } else {
+          changedParameters.push(parameter);
+        }
+      }
+    }
+
+    // create ParameterEvent
+    const parameterEvent = new(loader.loadInterface(PARAMETER_EVENT_MSG_TYPE));
+
+    const secondsAndNanos = (new Clock.ROSClock()).now().secondsAndNanoseconds;
+    parameterEvent.stamp = { sec: secondsAndNanos.seconds, nanosec: secondsAndNanos.nanoseconds };
+
+    parameterEvent.node = this.namespace() === '/' ?
+      this.namespace() + this.name() :
+      this.namespace() + '/' + this.name();
+
+    if (newParameters.length > 0) {
+      // eslint-disable-next-line camelcase
+      parameterEvent.new_parameters =
```

Comment:
You can avoid needing to disable eslint rules by using the [] for property assignments. E.g. `parameterEvent['new_parameter'] = ...`. If you prefer to stick with what you currently have that's fine with me too.

Reply:
I can go either way. Sticking with disabling eslint rule atm to get this merge done. Will discuss w/ you and @minggangw later and will gladly revise eslint rule disabling is discouraged for such cases.